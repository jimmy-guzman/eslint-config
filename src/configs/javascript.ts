import type { TypedConfigItem } from "../types";

import { GLOB_TESTS } from "../globs";
import { javascriptRules } from "../rules/javascript";

export const javascriptConfig = () => {
  return [
    {
      linterOptions: {
        reportUnusedDisableDirectives: true,
      },
      name: "jimmy.codes/javascript",
      rules: javascriptRules,
    },
    {
      files: GLOB_TESTS,
      name: "jimmy.codes/javascript/testing",
      rules: {
        "no-magic-numbers": "off",
      },
    },
  ] satisfies TypedConfigItem[];
};
