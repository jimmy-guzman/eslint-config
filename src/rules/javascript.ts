import eslint from "@eslint/js";

import type { Rules } from "../types";

export const additionalRules = {
  "array-callback-return": [
    "error",
    {
      allowImplicit: true, // Allow implicit return for flexibility while enforcing return consistency
    },
  ],
  "arrow-body-style": ["error", "always"], // Enforce `{}` in arrow functions for consistency (open to revisiting)
  "class-methods-use-this": "error", // Ensures class methods are used within their class (avoids unnecessary methods)
  "consistent-return": "error", // Prevents inconsistent function returns (e.g., sometimes returning `undefined`)
  "curly": ["error", "all"], // Enforce `{}` around all control statements (avoids accidental bugs)
  "default-case": "error", // Require `default` case in switch statements (prevents unintended fallthrough)
  "default-case-last": "error", // Ensure the `default` case is the last one in `switch`
  "eqeqeq": "error", // Enforce strict equality (`===` and `!==`) over abstract equality (`==` and `!=`)
  "no-console": "warn", // Allow logging in development, but should be reviewed for production builds
  "no-div-regex": "error",
  "no-else-return": "error", // Disallow `else` blocks after `return` statements in `if` blocks
  "no-implicit-coercion": "error", // Prevents shorthand type conversions (e.g., `!!value`, `+var`)
  "no-implicit-globals": "error", // Prevents defining variables in the global scope
  "no-lonely-if": "error", // Disallow `if` statements that are the only statement in a block
  "no-loop-func": "error", // Disallow function declarations inside loops (prevents scoping issues)
  "no-magic-numbers": ["error", { ignore: [0, 1, -1, 2] }], // Helps catch unexplained numbers while allowing common values
  "no-new-wrappers": "error", // Prevents `new String()`, `new Number()`, etc. (use literals instead)
  "no-param-reassign": ["error", { props: true }], // Disallow reassigning function parameters but allow modifying object properties
  "no-promise-executor-return": "error", // Ensures `new Promise` executors always return something
  "no-self-compare": "error", // Disallows `x === x` (usually a mistake)
  "no-template-curly-in-string": "error", // Flags unintended use of `${}` inside strings without backticks
  "no-throw-literal": "error", // Prevents throwing non-error objects (e.g., `throw "error"` should be `throw new Error("error")`)
  "no-unassigned-vars": "error", // Disallow let or var variables that are read but never assigned
  "no-unmodified-loop-condition": "error", // Ensures loop conditions change, preventing infinite loops
  "no-unneeded-ternary": "error", // Disallow ternary operators when simpler alternatives exist
  "no-unreachable-loop": "error", // Prevents loops that can never execute
  "no-use-before-define": [
    "error",
    {
      allowNamedExports: false,
      classes: false,
      functions: true, // Allow function hoisting
      variables: true,
    },
  ],
  "no-useless-computed-key": "error", // Disallows unnecessary computed keys in objects (e.g., `{["key"]: value}` instead of `{ key: value }`)
  "no-useless-constructor": "error", // Disallows empty constructors
  "no-useless-rename": "error", // Disallows renaming variables to the same name in import/export/destructuring
  "no-useless-return": "error", // Prevents redundant `return;` statements
  "no-var": "error", // Enforce `let` and `const` (modern best practice)
  "object-shorthand": "error", // Enforces `{ foo }` instead of `{ foo: foo }`
  "prefer-arrow-callback": "error", // Enforces arrow functions in callbacks where possible
  "prefer-const": "error", // Requires `const` when a variable isnâ€™t reassigned
  "prefer-destructuring": [
    "error",
    {
      AssignmentExpression: {
        array: false, // Allow `x = arr[0]`, don't force `[x] = arr`
        object: false, // Allow `x = obj.x`, don't force `{ x } = obj`
      },
      VariableDeclarator: {
        array: false,
        object: true, // Enforce destructuring for objects (`const { x } = obj`)
      },
    },
  ],
  "prefer-object-spread": "error", // Enforces `{ ...obj }` over `Object.assign()`
  "prefer-rest-params": "error", // Use `...args` instead of `arguments`
  "prefer-spread": "error", // Use `[...arr]` instead of `arr.concat()`
  "prefer-template": "error", // Use template literals instead of string concatenation
  "radix": "error", // Require a radix argument in `parseInt`
  "require-await": "error", // Disallows async functions without `await` (prevents unnecessary `async`)
  "strict": ["error", "safe"], // Enforces `"use strict"` only where needed (safe mode)
  "symbol-description": "error", // Requires descriptions when creating `Symbol()`
} satisfies Rules;

export const javascriptRules = {
  ...eslint.configs.recommended.rules,
  ...additionalRules,
} satisfies Rules;
