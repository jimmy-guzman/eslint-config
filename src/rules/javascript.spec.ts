import { javascriptRules } from "./javascript";

test("should create javascript rules", () => {
  expect(javascriptRules).toMatchSnapshot();
});
