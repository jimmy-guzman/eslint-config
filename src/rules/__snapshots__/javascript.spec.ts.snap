// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should create javascript rules 1`] = `
{
  "array-callback-return": [
    "error",
    {
      "allowImplicit": true,
    },
  ],
  "arrow-body-style": [
    "error",
    "always",
  ],
  "class-methods-use-this": "error",
  "consistent-return": "error",
  "constructor-super": "error",
  "curly": [
    "error",
    "all",
  ],
  "default-case": "error",
  "default-case-last": "error",
  "eqeqeq": "error",
  "for-direction": "error",
  "getter-return": "error",
  "no-async-promise-executor": "error",
  "no-case-declarations": "error",
  "no-class-assign": "error",
  "no-compare-neg-zero": "error",
  "no-cond-assign": "error",
  "no-console": "warn",
  "no-const-assign": "error",
  "no-constant-binary-expression": "error",
  "no-constant-condition": "error",
  "no-control-regex": "error",
  "no-debugger": "error",
  "no-delete-var": "error",
  "no-div-regex": "error",
  "no-dupe-args": "error",
  "no-dupe-class-members": "error",
  "no-dupe-else-if": "error",
  "no-dupe-keys": "error",
  "no-duplicate-case": "error",
  "no-else-return": "error",
  "no-empty": "error",
  "no-empty-character-class": "error",
  "no-empty-pattern": "error",
  "no-empty-static-block": "error",
  "no-ex-assign": "error",
  "no-extra-boolean-cast": "error",
  "no-fallthrough": "error",
  "no-func-assign": "error",
  "no-global-assign": "error",
  "no-implicit-coercion": "error",
  "no-implicit-globals": "error",
  "no-import-assign": "error",
  "no-invalid-regexp": "error",
  "no-irregular-whitespace": "error",
  "no-lonely-if": "error",
  "no-loop-func": "error",
  "no-loss-of-precision": "error",
  "no-magic-numbers": [
    "error",
    {
      "ignore": [
        0,
        1,
        -1,
        2,
      ],
    },
  ],
  "no-misleading-character-class": "error",
  "no-new-native-nonconstructor": "error",
  "no-new-wrappers": "error",
  "no-nonoctal-decimal-escape": "error",
  "no-obj-calls": "error",
  "no-octal": "error",
  "no-param-reassign": [
    "error",
    {
      "props": true,
    },
  ],
  "no-promise-executor-return": "error",
  "no-prototype-builtins": "error",
  "no-redeclare": "error",
  "no-regex-spaces": "error",
  "no-self-assign": "error",
  "no-self-compare": "error",
  "no-setter-return": "error",
  "no-shadow-restricted-names": "error",
  "no-sparse-arrays": "error",
  "no-template-curly-in-string": "error",
  "no-this-before-super": "error",
  "no-throw-literal": "error",
  "no-unassigned-vars": "error",
  "no-undef": "error",
  "no-unexpected-multiline": "error",
  "no-unmodified-loop-condition": "error",
  "no-unneeded-ternary": "error",
  "no-unreachable": "error",
  "no-unreachable-loop": "error",
  "no-unsafe-finally": "error",
  "no-unsafe-negation": "error",
  "no-unsafe-optional-chaining": "error",
  "no-unused-labels": "error",
  "no-unused-private-class-members": "error",
  "no-unused-vars": "error",
  "no-use-before-define": [
    "error",
    {
      "allowNamedExports": false,
      "classes": false,
      "functions": true,
      "variables": true,
    },
  ],
  "no-useless-backreference": "error",
  "no-useless-catch": "error",
  "no-useless-computed-key": "error",
  "no-useless-constructor": "error",
  "no-useless-escape": "error",
  "no-useless-rename": "error",
  "no-useless-return": "error",
  "no-var": "error",
  "no-with": "error",
  "object-shorthand": "error",
  "prefer-arrow-callback": "error",
  "prefer-const": "error",
  "prefer-destructuring": [
    "error",
    {
      "AssignmentExpression": {
        "array": false,
        "object": false,
      },
      "VariableDeclarator": {
        "array": false,
        "object": true,
      },
    },
  ],
  "prefer-object-spread": "error",
  "prefer-rest-params": "error",
  "prefer-spread": "error",
  "prefer-template": "error",
  "radix": "error",
  "require-await": "error",
  "require-yield": "error",
  "strict": [
    "error",
    "safe",
  ],
  "symbol-description": "error",
  "use-isnan": "error",
  "valid-typeof": "error",
}
`;
